enhancements/installer/component-selection.md

```diff
@@ -0,0 +1,358 @@
+clusters to the components I need for the service.
+
+* As a user creating a cluster that will never run an image registry, I do not want the additional overhead
+of running the image registry operator, or have to remove the default registry that is created.
+
+* As a team packaging openshift for a specific use case such as edge deployments, I want to provide
+an install experience that disables components that aren't needed for my use case.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The CVO already has the ability to respect annotations on resources, as can be seen
+[here](https://github.com/openshift/cluster-kube-apiserver-operator/blob/c03c9edf5fddf4e3fb1bc6d7afcd2a2284ca03d8/manifests/0000_20_kube-apiserver-operator_06_deployment.yaml#L10) and leveraged [here](https://github.com/openshift/hypershift/blob/main/control-plane-operator/controllers/hostedcontrolplane/assets/cluster-version-operator/cluster-version-operator-deployment.yaml#L47-L48).
+This proposal consists of two parts:
+
+1) Formalizing a concept of a "capability" annotation which allows a given resource to be excluded based
+on installer input. For example the console related resources could be annotated as
+
+```yaml
+annotations:
+  capability.openshift.io/console: "true"
+```
+
+2) Defining an [install config api](https://github.com/openshift/installer/blob/790048166067273d34f76bea4220fa395b1cce1b/pkg/types/installconfig.go#L70) field whereby the user can opt out of specific capabilities.
+
+InstallConfig.ExcludeCapabilities
+- console
+- samples
+
+Which resources ultimately get installed for a given cluster would be the set of resources encompassed
+by the CLUSTER_PROFILE(if any), minus any resources explicitly excluded by the excluded capabilities configuration.
+
+Examples of candidate components to be treated as capabilities:
+
+* console
+* imageregistry
+* samples
+* cluster baremetal operator
+* olm/marketplace
+* kube-storage-version-migrator
+* csi-*
+* insights
+* monitoring
+* ???
+
+To distinguish between the user intent of "Give me the default components that are suitable for my install" and
+"do not exclude any components", users can specify `none` as the capability to exclude.  In the event that we
+want the installer to apply some default exclusions in specific scenarios when the user has not expressed any
+specific desire to exclude particular components, this `none` specification will be a way for the user to opt out
+of having those defaults applied.
+
+Alternative:
+If we think it's better to invert it, such that "if you tell us nothing, you get all the components", we can instead
+define a `default` component keyword, such that if you don't specify any exclusions, the installer leaves it alone
+and you get everything, but if you specify `default`, then the installer knows to replace the exclusion list with
+whatever the installer thinks is best for your particular platform/whatever.
```

Comment:
hm.  i'm willing to add the "action" field to the api to give us that futureproofing, but i'd prefer not to define "include" as a valid action yet (so initially the only valid action would be "exclude").

For the include case, i can see solving it two other ways, thus i'm not sure we'd need it:
1) move the component to olm, if it's really a thing that most installs(both as measured by instances of installs, but also types of installs like sno vs hypershift vs osd vs standalone, etc) do not want turned on
2) embed the "disabled by default" list in the installer (so if you want those components, you pass the `none` filterlist, or the filterlist that only filters the things you really don't want).

wdyt?

Reply:
> I guess for your option 2, if the user specifies anything in the capabilities list then the defaults are ignored and we do what they say? 

correct.

> So if there are 2 things disabled by default and the user specifies a different third item but those 2, they are installed? I think that could be surprising.

could be, but we can document "here are all the things that are disabled by default, if you want to disable things, you need to list these also".  Though it does not solve the upgrade case (you upgrade to a version that has a new capability, which is disabled by default, but now your cluster will get it).

So in short i think we can:

1) add the action field (default action is exclude, or just require it but exclude is the only valid action)
2) use "true" to annotate things that are included, but can be excluded
3) someday, we can introduce an "include" action, and allow "false" to mean "excluded but can be included"

Reply:
If we use separate lists we have to declare which takes precedence in the case someone messes up and lists a component in both. With 1 list we can say "last instruction wins". If we have 2, maybe `exclude` overrides `include` because the user can undo the exclusion in the cluster if that was a mistake?

Reply:
> With 1 list we can say "last instruction wins".

isn't that also defining a precedence?

if you explicitly told us to both "include" something and "exclude" something, that can be an error condition, too.